docs/eh_v3.md

```diff
@@ -0,0 +1,140 @@
+
+To start listening EHorizon updates `EHorizonObserver` should be registered with
+
+```kotlin
+mapboxNavigation.registerEHorizonObserver(eHorizonObserver)
+```
+
+It provides the next callbacks:
+
+```kotlin
+/**
+ * Electronic horizon listener. Callbacks are fired in the order specified.
+ * onPositionUpdated might be called multiple times after the other callbacks until a new change to
+ * the horizon occurs.
+ */
+interface EHorizonObserver {
+
+    /**
+     * This callback might be called multiple times when the position changes.
+     * @param position current electronic horizon position(map matched position + e-horizon tree)
+     * @param distances map road object id -> EHorizonObjectDistanceInfo for upcoming road objects
+     *
+     */
+    fun onPositionUpdated(
+        position: EHorizonPosition,
+        distances: Map<String, EHorizonObjectDistanceInfo>
+    )
+
+    /**
+     * Called when entry to line-like(i.e. which has length != null) road object was detected
+     * @param objectEnterExitInfo contains info related to the object
+     */
+    fun onRoadObjectEnter(objectEnterExitInfo: EHorizonObjectEnterExitInfo)
+
+    /**
+     * Called when exit from line-like(i.e. which has length != null) road object was detected
+     * @param objectEnterExitInfo contains info related to the object
+     */
+    fun onRoadObjectExit(objectEnterExitInfo: EHorizonObjectEnterExitInfo)
+
+    /**
+     * This callback is fired whenever road object is added
+     * @param roadObjectId id of the object
+     */
+    fun onRoadObjectAdded(roadObjectId: String)
+
+    /**
+     * This callback is fired whenever road object is updated
+     * @param roadObjectId id of the object
+     */
+    fun onRoadObjectUpdated(roadObjectId: String)
+
+    /**
+     * This callback is fired whenever road object is removed
+     * @param roadObjectId id of the object
+     */
+    fun onRoadObjectRemoved(roadObjectId: String)
+}
+```
+
+Anytime you need edge's shape or metadata you can get it with `EHorizonGraphAccessor`.
+It's available with
+```kotlin
+mapboxNavigation.getEHorizonGraphAccessor()
+```
+
+```kotlin
+interface EHorizonGraphAccessor {
+    /**
+     * Returns Graph Edge geometry for the given GraphId of the edge.
+     * If edge with given edgeId is not accessible, returns null
+     * @param edgeId
+     *
+     * @return list of Points representing edge shape
+     */
+    fun getEdgeShape(edgeId: Long): List<Point>?
+
+    /**
+     * Returns Graph Edge meta-information for the given GraphId of the edge.
+     * If edge with given edgeId is not accessible, returns null
+     * @param edgeId
+     *
+     * @return EHorizonEdgeMetadata
+     */
+    fun getEdgeMetadata(edgeId: Long): EHorizonEdgeMetadata?
+}
+```
+
+Anytime you need road object's metadata or location you can get it with `EHorizonObjectsStore`
+It's available with
+```kotlin
+mapboxNavigation.getEHorizonObjectsStore()
+```
+
+```kotlin
+interface EHorizonObjectsStore {
```

Comment:
Not sure about `EHorizon` prefix here. Road object store can be used not only for EH.

Reply:
@etl wdyt?

Reply:
RoadObjectStore seems like a more appropriate name, isn't it?

Reply:
In automotive world where EH was born, anything that supplies you with the data from underneath nav graph is called Electronic Horizon. So I see logic for this to be with this scoping prefix. However there might be different reason for making EH prefix, so curious to know.